File: backend/app/services/live_ltp_island.py
"""Closed grammar for Adaptive live LTP islands (canvas + copy flatten).

Keep in sync with frontend/lib/live-ltp-island.ts.
"""

from __future__ import annotations

import html
import re
from typing import Any

LIVE_LTP_MAX_ISLANDS = 24

_TOKEN_RE = re.compile(
    r"\{\{ltp:([A-Za-z0-9]+):([A-Za-z0-9._-]+)((?:\|[A-Za-z][A-Za-z0-9_]*=[^|}]+)+)?\}\}"
)
_TAG_RE = re.compile(r"<ananta-ltp\b([^>]*)(?:\/>|><\/ananta-ltp>|>)", re.I)
_SAFE_ATTR_KEYS = frozenset({"asOf", "chgPct", "kind", "ltp"})


def _parse_number(raw: str | None) -> float | None:
    if raw is None or raw == "":
        return None
    try:
        value = float(raw)
    except (TypeError, ValueError):
        return None
    return value if value == value else None  # NaN check


def _parse_kind(raw: str | None) -> str:
    value = str(raw or "").strip().lower()
    if value in {"chgpct", "chg", "pct"}:
        return "chgPct"
    if value in {"ltp", "price"}:
        return "ltp"
    return "both"


def island_key(exchange: str, symbol: str) -> str:
    return f"{(exchange or 'NSE').strip().upper()}:{(symbol or '').strip().upper()}"


def _attrs_from_parts(exchange_raw: str, symbol_raw: str, pipe_attrs: str) -> dict[str, Any] | None:
    exchange = str(exchange_raw or "").strip().upper()
    symbol = str(symbol_raw or "").strip().upper()
    if not exchange or not symbol:
        return None
    if not re.fullmatch(r"[A-Z0-9]+", exchange) or not re.fullmatch(r"[A-Z0-9._-]+", symbol):
        return None
    bag: dict[str, str] = {}
    for part in (pipe_attrs or "").split("|"):
        if not part or "=" not in part:
            continue
        key, value = part.split("=", 1)
        key = key.strip()
        value = value.strip()
        if key not in _SAFE_ATTR_KEYS:
            continue
        if re.search(r"[<>\"'`]", value) or re.search(r"\bon\w+", value, re.I):
            continue
        bag[key] = value
    return {
        "asOf": bag.get("asOf"),
        "chgPct": _parse_number(bag.get("chgPct")),
        "exchange": exchange,
        "kind": _parse_kind(bag.get("kind")),
        "ltp": _parse_number(bag.get("ltp")),
        "symbol": symbol,
    }


def parse_live_ltp_token(token: str) -> dict[str, Any] | None:
    match = _TOKEN_RE.fullmatch(token.strip())
    if not match:
        return None
    return _attrs_from_parts(match.group(1), match.group(2), match.group(3) or "")


def parse_ananta_ltp_attribute_string(raw: str) -> dict[str, Any] | None:
    def get(name: str) -> str | None:
        match = re.search(rf'\bdata-{name}\s*=\s*["\']([^"\']*)["\']', raw or "", re.I)
        return match.group(1) if match else None

    symbol = (get("symbol") or "").strip().upper()
    exchange = (get("exchange") or "NSE").strip().upper()
    if not symbol or not re.fullmatch(r"[A-Z0-9._-]+", symbol) or not re.fullmatch(r"[A-Z0-9]+", exchange):
        return None
    return {
        "asOf": get("as-of"),
        "chgPct": _parse_number(get("chg-pct")),
        "exchange": exchange,
        "kind": _parse_kind(get("kind")),
        "ltp": _parse_number(get("ltp")),
        "symbol": symbol,
    }


def format_live_ltp_flatten(symbol: str, displayed: dict[str, float | None], kind: str = "both") -> str:
    sym = (symbol or "").strip().upper() or "—"
    ltp = displayed.get("ltp")
    chg = displayed.get("chgPct")
    ltp_text = f"{ltp:,.2f}" if ltp is not None else None
    chg_text = None
    if chg is not None:
        sign = "+" if chg >= 0 else ""
        chg_text = f"{sign}{chg:.2f}%"
    if kind == "ltp":
        return f"{sym} {ltp_text}" if ltp_text else sym
    if kind == "chgPct":
        return f"{sym} ({chg_text})" if chg_text else sym
    if ltp_text and chg_text:
        return f"{sym} {ltp_text} ({chg_text})"
    if ltp_text:
        return f"{sym} {ltp_text}"
    if chg_text:
        return f"{sym} ({chg_text})"
    return f"{sym} —"


def serialize_ananta_ltp_element(attrs: dict[str, Any]) -> str:
    parts = [
        f'data-symbol="{html.escape(str(attrs["symbol"]), quote=True)}"',
        f'data-exchange="{html.escape(str(attrs["exchange"]), quote=True)}"',
        f'data-kind="{html.escape(str(attrs.get("kind") or "both"), quote=True)}"',
    ]
    if attrs.get("ltp") is not None:
        parts.append(f'data-ltp="{attrs["ltp"]}"')
    if attrs.get("chgPct") is not None:
        parts.append(f'data-chg-pct="{attrs["chgPct"]}"')
    if attrs.get("asOf"):
        parts.append(f'data-as-of="{html.escape(str(attrs["asOf"]), quote=True)}"')
    return f'<ananta-ltp {" ".join(parts)}></ananta-ltp>'


def flatten_live_ltp_islands(
    text: str,
    displayed: dict[str, dict[str, float | None]] | None = None,
    max_islands: int = LIVE_LTP_MAX_ISLANDS,
) -> str:
    displayed = displayed or {}
    count = 0

    def repl(match: re.Match[str]) -> str:
        nonlocal count
        attrs = _attrs_from_parts(match.group(1), match.group(2), match.group(3) or "")
        if not attrs:
            return match.group(0)
        count += 1
        key = island_key(attrs["exchange"], attrs["symbol"])
        live = displayed.get(key) or {}
        values = {
            "ltp": live.get("ltp", attrs.get("ltp")),
            "chgPct": live.get("chgPct", attrs.get("chgPct")),
        }
        return format_live_ltp_flatten(attrs["symbol"], values, attrs.get("kind") or "both")

    return _TOKEN_RE.sub(repl, text)


def live_ltp_tokens_to_elements(text: str, max_islands: int = LIVE_LTP_MAX_ISLANDS) -> str:
    count = 0

    def repl(match: re.Match[str]) -> str:
        nonlocal count
        attrs = _attrs_from_parts(match.group(1), match.group(2), match.group(3) or "")
        if not attrs:
            return match.group(0)
        count += 1
        if count > max_islands:
            return format_live_ltp_flatten(
                attrs["symbol"],
                {"ltp": attrs.get("ltp"), "chgPct": attrs.get("chgPct")},
                attrs.get("kind") or "both",
            )
        return serialize_ananta_ltp_element(attrs)

    return _TOKEN_RE.sub(repl, text)


def sanitize_ananta_ltp_elements(html_text: str, max_islands: int = LIVE_LTP_MAX_ISLANDS) -> str:
    count = 0

    def repl(match: re.Match[str]) -> str:
        nonlocal count
        attrs = parse_ananta_ltp_attribute_string(match.group(1) or "")
        if not attrs:
            return ""
        count += 1
        if count > max_islands:
            return format_live_ltp_flatten(
                attrs["symbol"],
                {"ltp": attrs.get("ltp"), "chgPct": attrs.get("chgPct")},
                attrs.get("kind") or "both",
            )
        return serialize_ananta_ltp_element(attrs)

    return _TAG_RE.sub(repl, html_text)


def prepare_live_ltp_in_html(document: str, max_islands: int = LIVE_LTP_MAX_ISLANDS) -> str:
    """Convert tokens then rewrite any ananta-ltp tags to allowlisted attrs only."""
    text = live_ltp_tokens_to_elements(document or "", max_islands=max_islands)
    return sanitize_ananta_ltp_elements(text, max_islands=max_islands)


def extract_live_ltp_symbols(document: str, max_islands: int = LIVE_LTP_MAX_ISLANDS) -> list[dict[str, str]]:
    prepared = prepare_live_ltp_in_html(document, max_islands=max_islands)
    out: list[dict[str, str]] = []
    seen: set[str] = set()
    for match in _TAG_RE.finditer(prepared):
        attrs = parse_ananta_ltp_attribute_string(match.group(1) or "")
        if not attrs:
            continue
        key = island_key(attrs["exchange"], attrs["symbol"])
        if key in seen:
            continue
        seen.add(key)
        out.append({"exchange": attrs["exchange"], "symbol": attrs["symbol"]})
        if len(out) >= max_islands:
            break
    return out

File: frontend/lib/live-ltp-island.ts
/** Closed grammar for Adaptive live LTP islands. Shareable with OSS. */

export const LIVE_LTP_MAX_ISLANDS = 24;
export const ANANTA_LTP_PROTOCOL = "ananta-ltp:";
/** Streamdown-safe path prefix (custom schemes are rewritten to [blocked]). */
export const ANANTA_LTP_PATH_PREFIX = "/__ananta-ltp__/";

export type LiveLtpKind = "ltp" | "chgPct" | "both";

export type LiveLtpIslandAttrs = {
    asOf?: string | null;
    chgPct?: number | null;
    exchange: string;
    kind: LiveLtpKind;
    ltp?: number | null;
    symbol: string;
};

export type LiveLtpDisplayed = {
    chgPct: number | null;
    ltp: number | null;
};

const TOKEN_RE =
    /\{\{ltp:([A-Za-z0-9]+):([A-Za-z0-9._-]+)((?:\|[A-Za-z][A-Za-z0-9_]*=[^|}]+)+)?\}\}/g;

const SAFE_ATTR_KEYS = new Set(["asOf", "chgPct", "kind", "ltp"]);

function parseNumber(raw: string | undefined): number | null {
    if (raw == null || raw === "") return null;
    const n = Number(raw);
    return Number.isFinite(n) ? n : null;
}

function parseKind(raw: string | undefined): LiveLtpKind {
    const v = String(raw || "")
        .trim()
        .toLowerCase();
    if (v === "chgpct" || v === "chg" || v === "pct") return "chgPct";
    if (v === "ltp" || v === "price") return "ltp";
    return "both";
}

export function islandKey(exchange: string, symbol: string): string {
    return `${String(exchange || "NSE").trim().toUpperCase()}:${String(symbol || "")
        .trim()
        .toUpperCase()}`;
}

export function parseLiveLtpToken(token: string): LiveLtpIslandAttrs | null {
    const match = TOKEN_RE.exec(token);
    TOKEN_RE.lastIndex = 0;
    if (!match || match[0] !== token.trim()) return null;
    return attrsFromMatch(match[1], match[2], match[3] || "");
}

function attrsFromMatch(exchangeRaw: string, symbolRaw: string, pipeAttrs: string): LiveLtpIslandAttrs | null {
    const exchange = String(exchangeRaw || "")
        .trim()
        .toUpperCase();
    const symbol = String(symbolRaw || "")
        .trim()
        .toUpperCase();
    if (!exchange || !symbol || !/^[A-Z0-9]+$/.test(exchange) || !/^[A-Z0-9._-]+$/.test(symbol)) {
        return null;
    }
    const bag: Record<string, string> = {};
    for (const part of pipeAttrs.split("|")) {
        if (!part) continue;
        const eq = part.indexOf("=");
        if (eq <= 0) continue;
        const key = part.slice(0, eq).trim();
        const value = part.slice(eq + 1).trim();
        if (!SAFE_ATTR_KEYS.has(key)) continue;
        if (/[<>"'`]/.test(value) || /\bon\w+/i.test(value)) continue;
        bag[key] = value;
    }
    return {
        asOf: bag.asOf || null,
        chgPct: parseNumber(bag.chgPct),
        exchange,
        kind: parseKind(bag.kind),
        ltp: parseNumber(bag.ltp),
        symbol
    };
}

export function parseLiveLtpAttrsFromMatch(
    exchange: string,
    symbol: string,
    pipeAttrs: string
): LiveLtpIslandAttrs | null {
    return attrsFromMatch(exchange, symbol, pipeAttrs);
}

/** Extract islands from markdown / HTML without mutating. Caps at max. */
export function extractLiveLtpIslands(text: string, max = LIVE_LTP_MAX_ISLANDS): LiveLtpIslandAttrs[] {
    const out: LiveLtpIslandAttrs[] = [];
    const seen = new Set<string>();
    TOKEN_RE.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = TOKEN_RE.exec(text)) && out.length < max) {
        const attrs = attrsFromMatch(match[1], match[2], match[3] || "");
        if (!attrs) continue;
        const key = islandKey(attrs.exchange, attrs.symbol);
        if (seen.has(key)) continue;
        seen.add(key);
        out.push(attrs);
    }
    TOKEN_RE.lastIndex = 0;

    const tagRe =
        /<ananta-ltp\b([^>]*)\/?>/gi;
    let tagMatch: RegExpExecArray | null;
    while ((tagMatch = tagRe.exec(text)) && out.length < max) {
        const attrs = parseAnantaLtpAttributeString(tagMatch[1] || "");
        if (!attrs) continue;
        const key = islandKey(attrs.exchange, attrs.symbol);
        if (seen.has(key)) continue;
        seen.add(key);
        out.push(attrs);
    }
    return out;
}

export function parseAnantaLtpAttributeString(raw: string): LiveLtpIslandAttrs | null {
    if (/\bon\w+\s*=/i.test(raw) || /\b(?:href|src)\s*=/i.test(raw)) {
        // Still allow the element if we can recover safe data-* only — strip danger by ignoring those attrs.
    }
    const get = (name: string): string | null => {
        const re = new RegExp(`\\bdata-${name}\\s*=\\s*["']([^"']*)["']`, "i");
        const m = re.exec(raw);
        return m ? m[1] : null;
    };
    const symbol = (get("symbol") || "").trim().toUpperCase();
    const exchange = (get("exchange") || "NSE").trim().toUpperCase();
    if (!symbol || !/^[A-Z0-9._-]+$/.test(symbol) || !/^[A-Z0-9]+$/.test(exchange)) return null;
    return {
        asOf: get("as-of"),
        chgPct: parseNumber(get("chg-pct") || undefined),
        exchange,
        kind: parseKind(get("kind") || undefined),
        ltp: parseNumber(get("ltp") || undefined),
        symbol
    };
}

export function formatLiveLtpFlatten(
    symbol: string,
    displayed: LiveLtpDisplayed,
    kind: LiveLtpKind = "both"
): string {
    const sym = symbol.trim().toUpperCase() || "—";
    const ltp =
        displayed.ltp != null && Number.isFinite(displayed.ltp)
            ? new Intl.NumberFormat("en-IN", { maximumFractionDigits: 2 }).format(displayed.ltp)
            : null;
    const chg =
        displayed.chgPct != null && Number.isFinite(displayed.chgPct)
            ? `${displayed.chgPct >= 0 ? "+" : ""}${displayed.chgPct.toFixed(2)}%`
            : null;
    if (kind === "ltp") {
        return ltp != null ? `${sym} ${ltp}` : sym;
    }
    if (kind === "chgPct") {
        return chg != null ? `${sym} (${chg})` : sym;
    }
    if (ltp != null && chg != null) return `${sym} ${ltp} (${chg})`;
    if (ltp != null) return `${sym} ${ltp}`;
    if (chg != null) return `${sym} (${chg})`;
    return `${sym} —`;
}

/**
 * Replace island tokens (and ananta-ltp markdown links) with clipboard-friendly text.
 * Prefer `displayed` map (live values) over snapshot attrs in the token.
 */
export function flattenLiveLtpIslands(
    text: string,
    displayed: Record<string, LiveLtpDisplayed> = {},
    max = LIVE_LTP_MAX_ISLANDS
): string {
    let count = 0;
    let out = text.replace(TOKEN_RE, (full, exchange: string, symbol: string, pipeAttrs: string) => {
        const attrs = attrsFromMatch(exchange, symbol, pipeAttrs || "");
        if (!attrs) return full;
        count += 1;
        if (count > max) {
            return formatLiveLtpFlatten(attrs.symbol, { ltp: attrs.ltp ?? null, chgPct: attrs.chgPct ?? null }, attrs.kind);
        }
        const key = islandKey(attrs.exchange, attrs.symbol);
        const live = displayed[key];
        return formatLiveLtpFlatten(
            attrs.symbol,
            {
                chgPct: live?.chgPct ?? attrs.chgPct ?? null,
                ltp: live?.ltp ?? attrs.ltp ?? null
            },
            attrs.kind
        );
    });

    out = out.replace(
        /\[([^\]]*)\]\(\/__ananta-ltp__\/([^/?#]+)\/([^)?#]+)(\?[^)]*)?\)/gi,
        (_full, _label: string, exchange: string, symbol: string, query = "") => {
            const params = new URLSearchParams(String(query || "").replace(/^\?/, ""));
            const attrs: LiveLtpIslandAttrs = {
                asOf: params.get("asOf"),
                chgPct: parseNumber(params.get("chgPct") || undefined),
                exchange: String(exchange || "NSE").toUpperCase(),
                kind: parseKind(params.get("kind") || undefined),
                ltp: parseNumber(params.get("ltp") || undefined),
                symbol: String(symbol || "").toUpperCase()
            };
            const key = islandKey(attrs.exchange, attrs.symbol);
            const live = displayed[key];
            return formatLiveLtpFlatten(
                attrs.symbol,
                {
                    chgPct: live?.chgPct ?? attrs.chgPct ?? null,
                    ltp: live?.ltp ?? attrs.ltp ?? null
                },
                attrs.kind
            );
        }
    );
    // Legacy ananta-ltp:// links (pre-sanitize path) still flatten if present in clipboard source.
    out = out.replace(
        /\[([^\]]*)\]\(ananta-ltp:\/\/([^/?#]+)\/([^)?#]+)(\?[^)]*)?\)/gi,
        (_full, _label: string, exchange: string, symbol: string, query = "") => {
            const params = new URLSearchParams(String(query || "").replace(/^\?/, ""));
            const attrs: LiveLtpIslandAttrs = {
                asOf: params.get("asOf"),
                chgPct: parseNumber(params.get("chgPct") || undefined),
                exchange: String(exchange || "NSE").toUpperCase(),
                kind: parseKind(params.get("kind") || undefined),
                ltp: parseNumber(params.get("ltp") || undefined),
                symbol: String(symbol || "").toUpperCase()
            };
            const key = islandKey(attrs.exchange, attrs.symbol);
            const live = displayed[key];
            return formatLiveLtpFlatten(
                attrs.symbol,
                {
                    chgPct: live?.chgPct ?? attrs.chgPct ?? null,
                    ltp: live?.ltp ?? attrs.ltp ?? null
                },
                attrs.kind
            );
        }
    );
    return out;
}

/** Preprocess agent tokens into Streamdown-safe markdown links (link-hijack path). */
export function liveLtpTokensToMarkdownLinks(text: string, max = LIVE_LTP_MAX_ISLANDS): string {
    let count = 0;
    return text.replace(TOKEN_RE, (full, exchange: string, symbol: string, pipeAttrs: string) => {
        const attrs = attrsFromMatch(exchange, symbol, pipeAttrs || "");
        if (!attrs) return full;
        count += 1;
        if (count > max) {
            return formatLiveLtpFlatten(attrs.symbol, { ltp: attrs.ltp ?? null, chgPct: attrs.chgPct ?? null }, attrs.kind);
        }
        const qs = new URLSearchParams();
        if (attrs.ltp != null) qs.set("ltp", String(attrs.ltp));
        if (attrs.chgPct != null) qs.set("chgPct", String(attrs.chgPct));
        if (attrs.asOf) qs.set("asOf", attrs.asOf);
        if (attrs.kind !== "both") qs.set("kind", attrs.kind);
        const q = qs.toString();
        return `[${attrs.symbol}](${ANANTA_LTP_PATH_PREFIX}${attrs.exchange}/${attrs.symbol}${q ? `?${q}` : ""})`;
    });
}

export function parseAnantaLtpHref(href: string | null | undefined): LiveLtpIslandAttrs | null {
    if (!href) return null;
    const trimmed = href.trim();
    if (trimmed.startsWith(ANANTA_LTP_PATH_PREFIX) || trimmed.startsWith("/__ananta-ltp__/")) {
        try {
            const url = new URL(trimmed, "https://ananta.local");
            const parts = url.pathname.replace(/^\/__ananta-ltp__\//, "").split("/");
            const exchange = (parts[0] || "NSE").toUpperCase();
            const symbol = decodeURIComponent(parts[1] || "").toUpperCase();
            if (!symbol) return null;
            return {
                asOf: url.searchParams.get("asOf"),
                chgPct: parseNumber(url.searchParams.get("chgPct") || undefined),
                exchange,
                kind: parseKind(url.searchParams.get("kind") || undefined),
                ltp: parseNumber(url.searchParams.get("ltp") || undefined),
                symbol
            };
        } catch {
            return null;
        }
    }
    if (!trimmed.toLowerCase().startsWith("ananta-ltp://")) return null;
    try {
        const url = new URL(trimmed);
        const exchange = (url.hostname || "NSE").toUpperCase();
        const symbol = decodeURIComponent(url.pathname.replace(/^\//, "")).toUpperCase();
        if (!symbol) return null;
        return {
            asOf: url.searchParams.get("asOf"),
            chgPct: parseNumber(url.searchParams.get("chgPct") || undefined),
            exchange,
            kind: parseKind(url.searchParams.get("kind") || undefined),
            ltp: parseNumber(url.searchParams.get("ltp") || undefined),
            symbol
        };
    } catch {
        return null;
    }
}

/** Convert tokens to safe <ananta-ltp> elements for canvas HTML. */
export function liveLtpTokensToElements(text: string, max = LIVE_LTP_MAX_ISLANDS): string {
    let count = 0;
    return text.replace(TOKEN_RE, (full, exchange: string, symbol: string, pipeAttrs: string) => {
        const attrs = attrsFromMatch(exchange, symbol, pipeAttrs || "");
        if (!attrs) return full;
        count += 1;
        if (count > max) {
            return formatLiveLtpFlatten(attrs.symbol, { ltp: attrs.ltp ?? null, chgPct: attrs.chgPct ?? null }, attrs.kind);
        }
        return serializeAnantaLtpElement(attrs);
    });
}

export function serializeAnantaLtpElement(attrs: LiveLtpIslandAttrs): string {
    const parts = [
        `data-symbol="${attrs.symbol}"`,
        `data-exchange="${attrs.exchange}"`,
        `data-kind="${attrs.kind}"`
    ];
    if (attrs.ltp != null) parts.push(`data-ltp="${attrs.ltp}"`);
    if (attrs.chgPct != null) parts.push(`data-chg-pct="${attrs.chgPct}"`);
    if (attrs.asOf) parts.push(`data-as-of="${attrs.asOf.replace(/"/g, "")}"`);
    return `<ananta-ltp ${parts.join(" ")}></ananta-ltp>`;
}

/** Rewrite any <ananta-ltp …> to allowlisted data-* only (strips onclick/src/href). */
export function sanitizeAnantaLtpElements(html: string, max = LIVE_LTP_MAX_ISLANDS): string {
    let count = 0;
    return html.replace(/<ananta-ltp\b([^>]*)(?:\/>|><\/ananta-ltp>|>)/gi, (full, attrRaw: string) => {
        const attrs = parseAnantaLtpAttributeString(attrRaw || "");
        if (!attrs) return "";
        count += 1;
        if (count > max) {
            return formatLiveLtpFlatten(attrs.symbol, { ltp: attrs.ltp ?? null, chgPct: attrs.chgPct ?? null }, attrs.kind);
        }
        return serializeAnantaLtpElement(attrs);
    });
}
